### DISCUSSION TITLE
Weird behavior when running compose from Visual Studio

### DISCUSSION THREAD

**Discussion by eff-vbov** [dotnet/dotnet-docker#5030] [Q&A] [answered]
Hey guys, I need help to understand something. 

I have a repository with a .NET service and a Dockerfile, and there is a Dockercompose file at the root of the repository. This compose can be run in Visual Studio. This service make use of NuGet packages that are stored in our company Azure DevOps Artifact registry and it need authentication. There is no Nuget.config in the repository, from visual studio we just have setup the nuget feed from the settings and the authentication is done via AD account, so no PAT needed.

The docker file is as follow:
 
```
FROM mcr.microsoft.com/dotnet/aspnet:6.0 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM mcr.microsoft.com/dotnet/aspnet:6.0 AS build
WORKDIR /src
COPY ["Libraries/TestApp.Domain/TestApp.Domain.csproj", "Services/TestApp.Domain/"]
COPY ["Services/TestApp.API/TestApp.API.csproj", "Services/TestApp.API/"]
RUN dotnet restore "Services/TestApp.API/TestApp.API.csproj"
COPY . .
WORKDIR "/src/Services/TestApp.API"
RUN dotnet build "TestApp.API.csproj" -c Debug -o /app

FROM build AS publish
RUN dotnet publish "TestApp.API.csproj" -c Debug -o /app

FROM base AS final
WORKDIR /app
COPY --from=publish /app .
ENTRYPOINT ["dotnet", "TestApp.API.dll"]
```

Now, with this Dockerfile I can run the docker compose from Visual Studio, image is build and started and everything is working as expected! HOW?
If I try to build the dockerfile from CLI outside VisualStudio, the image can't build because it cannot authenticate with the feed. (I mean, this is expected).

I am now trying to update the Dockerfile to use different base images:
```
FROM mcr.microsoft.com/dotnet/sdk:6.0-bookworm-slim AS build-publish
WORKDIR /app

COPY ["Libraries/TestApp.Domain/TestApp.Domain.csproj", "app/Libraries/TestApp.Domain/"]
COPY ["Services/TestApp.API/TestApp.API.csproj", "app/Services/TestApp.API/"]

RUN dotnet restore "app/Services/TestApp.API/TestApp.API.csproj"  --runtime linux-x64

RUN dotnet build "./src/Services/TestApp.API/TestApp.API.csproj" \
    -c Release \
	--runtime linux-x64 \
	--self-contained true \
	--no-restore

RUN dotnet publish "./src/Services/TestApp.API/TestApp.API.csproj" \
	-c Release \
	-o /app/publish \
	--no-restore \
	--no-build \
	--self-contained true \
	--runtime linux-x64

FROM mcr.microsoft.com/dotnet/runtime-deps:6.0-jammy-chiseled
WORKDIR /app

COPY --from=build-publish /app/publish .

ENTRYPOINT ["./TestApp.API"]
```

With this Dockerfile running the docker compose from VisualStudio is not able to build the image due to missing authentication to the feed, same as from building from CLI. (Kinda expected though).

Can someone please explain me what kind of black magic happens in Visual Studio when using the aspnet image? Because you can see both Dockerfile and I do not understand why with one image I can build without authenticating to the feed and the other not.

**Comment by mthalman** [comment, marked as answer]
Assuming your using [fast mode](https://learn.microsoft.com/en-us/visualstudio/containers/container-build?view=vs-2022#debugging), VS is actually building your project outside the context of Docker and just copying the output into the container that will be used for debugging. So in that case, it uses all of VS's authentication abilities for the NuGet feed since the restore happens outside of a container.

**Reply by eff-vbov** [reply]
Thank you very much for explaining! :) I wasn't aware about this feature, look like fast mode is enabled by default, unless specified otherwise in the project file.
